### PR TITLE
Introduce module `preprocessing` for distributed scaling/normalization 

### LINKED ISSUE #400: Implement baseline preprocessing module
**Related**
Most likely requires #399 to be finished first.

**Feature functionality**
Implement a baseline preprocessing module. Preprocessors are all kind of features scalars (standardizer, normalizer, min-max-scaler, ...). For the first module implementation iteration three preprocessors (the ones previously mentioned) are sufficient. Also expose the non-object oriented module-level functional style calls.

**Additional context**
Sklearn provides a well-designed preprocessor API. One might want to get inspired/adapt their design [https://scikit-learn.org/stable/modules/classes.html#module-sklearn.preprocessing](https://scikit-learn.org/stable/modules/classes.html#module-sklearn.preprocessing)

### PR DESCRIPTION
**Pull request by mrfh92**
Inteded to resolve #400: in this PR we want to implement a preprocessing module for heat that has roughly the same capabilities as [the one of scikit-learn](https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.StandardScaler.html). 

So far I have implemented: 

* StandardScaler 
* MinMaxScaler
* MaxAbsScaler
* RobustScaler
* Normalizer 

... 

**Idea:** use `heat.BaseEstimator` and keep as much of the scikit-learn API as possible

### CODE REVIEW

**Comment by mrfh92** [comment]
current problem in `StandardScaler`: sth is wrong with the variance computation...

**Comment by mrfh92** [comment]
@krajsek From your experience in the neuroscience community, is there another preprocessing tool (e.g., from the list of scikit-learn) we should definitively add?

**Comment by krajsek** [comment]
@mrfh92 No, not at that level. There are a large number of preprocessing steps used in neuroimaging pipelines but they are more own image processing algorithms like motion correction, slice timing corrction or registration. But on this level, simple normalization etc. I cannot think about any neuroscience specific preprocesssing steps. Apart from that I think OneHotEncoder is quite usefull.

**Comment by mrfh92** [comment]
to be discussed: 
* which additional preprocessing routines (e.g., from those available in `scikit-learn`) could/should be added?
* which of the routines from this module should be part of continuous benchmarking

**Comment by mrfh92** [comment]
extensive citation of scikit-learn should be added in the docs

**Comment by mrfh92** [comment]
@mtar I have addressed most issues you mentioned in the last PR talk 
only type hints for returned `self` seem to be a bit tricky, because `from typing import Self` produces an error for me...

**Comment by mtar** [comment]
@mrfh92 the _Self_ type was introduced in Python 3.11, on older versions you can import it from the _typing_extensions_ module

**Comment by mrfh92** [comment]
@mtar thanks for the hint ... i have added the respective changes

**Review by ClaudiaComito — CHANGES_REQUESTED** [review]
Thank you so much @mrfh92 , this is very timely.  I have a few comments.

File: heat/core/io.py [hunk 1]
```diff
@@ -108,6 +112,16 @@ def load_hdf5(
             raise TypeError("dataset must be str, not {}".format(type(dataset)))
         elif split is not None and not isinstance(split, int):
             raise TypeError(f"split must be None or int, not {type(split)}")
+        if load_fraction is not None:
+            if isinstance(load_fraction, float):
+                if load_fraction <= 0.0 or load_fraction >= 1.0:
+                    raise ValueError("load_fraction must be strictly between 0. and 1.")
+                if split is None:
+                    raise ValueError("load_fraction can only be used if split is not None.")
```
[the remark this answers is not in the document]
Comment (mrfh92) [reply heat/core/io.py:120]: I have chosen 2.

File: heat/preprocessing/preprocessing.py [hunk 2]
```diff
@@ -0,0 +1,612 @@
...
+        ----------
+        X : DNDarray of shape (n_datapoints, n_features)
+            The data set to be scaled.
+        """
+        _check_if_2D_float_DNDarray(X)
+        _check_n_features(self.scale_, X)
+        if self.copy:
+            Y = X * self.scale_
+            return Y
+        else:
```
[the remark this answers is not in the document]
Comment (mrfh92) [reply heat/preprocessing/preprocessing.py:430]: if #1184 has been done, this will make a difference i guess (?)

[the remark this answers is not in the document]
Comment (mrfh92) [reply heat/preprocessing/preprocessing.py:430]: sry misunderstood this ... you're right

File: heat/preprocessing/preprocessing.py [hunk 3]
```diff
@@ -0,0 +1,612 @@
...
+            _check_n_features(self.center_, X)
+        if self.with_scaling:
+            _check_n_features(self.scale_, X)
+        if self.copy:
+            Y = X
+            if self.with_centering:
+                Y -= self.center_
+            if self.with_scaling:
+                Y *= self.scale_
+            return Y
```
[the remark this answers is not in the document]
Comment (mrfh92) [reply heat/preprocessing/preprocessing.py:578]: I have added `.copy()` here. Full functionality of in-place transformations can be assessed only if #1184 has been implemented.

**Comment by mrfh92** [comment]
I have done the requested changes

**Review by ClaudiaComito — APPROVED** [review]
Thanks for the changes @mrfh92 !
